srcs/cgi/Cgi.cpp

```diff
@@ -0,0 +1,205 @@
+//
+// Created by yuumo on 2022/11/07.
+//
+
+/*
+ * According to RFC 3875, CGI responses can be document-response, local-redir-response, client-redir-response, client-redirdoc-response, but our implementation is limited to document-response.
+ */
+
+
+#include <cstdlib>
+#include <unistd.h>
+#include <cerrno>
+#include <vector>
+#include <cstring>
+#include <iostream>
+#include <cstdarg>
+#include <cstdio>
+#include <sys/socket.h>
+#include <cstdio>
+#include <sys/wait.h>
+#include "Cgi.hpp"
+
+Cgi::Cgi(ft::ServerChild server_child)
+    : request_method_(server_child.Get_HTTPHead().GetRequestMethod()), script_name_("hoge") {
+}
+
+Cgi::~Cgi() {
+  close(cgi_socket_);
+}
+
+/**
+ *
+ * @param request_method
+ *
+ * TODO: We must build this function
+ * - set environment
+ *
+ * https://datatracker.ietf.org/doc/html/rfc3875#section-4-1
+ * http://bashhp.web.fc2.com/WWW/header.html
+ */
+void Cgi::CreateEnvMap() {
+  cgi_env_val_["SERVER_SOFTWARE"] = "42";
+  cgi_env_val_["GATEWAY_INTERFACE"] = "CGI/1.1";
+  cgi_env_val_["SERVER_PROTOCOL"] = "HTTP/1.1"; // tmp
+  cgi_env_val_["SERVER_PORT"] = "80"; // tmp 不要かも
+  cgi_env_val_["SERVER_NAME"] = "webserv"; //tmp
+  cgi_env_val_["REQUEST_METHOD"] = request_method_;
+  cgi_env_val_["SCRIPT_NAME"] = script_name_; // tmp TODO: get_script_name()
+
+  // POST によってフォームを受信する場合に、標準入力から読み込む必要のあるbyte数
+  cgi_env_val_["CONTENT_LENGTH"] = "42"; // TODO: get_content_length()
+
+  // POST によってフォームを受信する場合
+  cgi_env_val_["CONTENT_TYPE"] = "html"; // tmp
+
+  // for GET
+  cgi_env_val_["QUERY_STRING"] = "query_string"; // tmp TODO: get_query_string()
+  cgi_env_val_["PATH_INFO"] = "";
+
+  cgi_env_val_["REQUEST_URI"] = ""; // 不要かも
+}
+
+/**
+ * @brief set environ based on cgi_env_val_
+ */
+void  Cgi::SetEnv() {
+  for (std::map<std::string, std::string>::const_iterator iter = cgi_env_val_.begin(); iter != cgi_env_val_.end(); ++iter) {
+    setenv(iter->first.c_str(), iter->second.c_str(), 1); // overwrite
+  }
+}
+
+/**
+ * @param from
+ * @param to
+ * @return
+ */
+int Cgi::change_fd(int from, int to) {
+  int ret_val;
+
+  if (from == to) {
+    return 0;
+  }
+  close(to);
+  ret_val = dup2(from, to);

```

Comment:
I need a more detailed explanation from you🙇🏻‍♂️

Reply:
了解です！すみません、何でも system callのreturn valueを絶対に確認したいと思っていました。
その通りに特にdup2()の関数は自動的に 'to' のfdをcloseしますので、83行目がいらないと思います。
close行目がなければ、確認する必要がないようになります。
でも後はchild processだけやっていると思いますので変更しなくて大丈夫と思います
意味はありますか？

Yes, sorry for the confusion! I was just thought it would be good to check every single system call.
the dup2 function will close 'to' fd, so we don't need the close function (and we don't need to check the result because we never call the function).
But I realized maybe it is only in the child process, so It is probably okay to not check it and not change it :)
Does it make sense?

Reply:
> If newfd was open, any errors that would have been reported at
       close(2) time are lost.  If this is of concern, then—unless the
       program is single-threaded and does not allocate file descriptors
       in signal handlers—the correct approach is not to close newfd
       before calling dup2(), because of the race condition described
       above.



https://man7.org/linux/man-pages/man2/dup.2.html